### PR TITLE
Generalized Linear Models (GLM)

### PR DESCRIPTION
**Pull request by VasanthakumarV**
Port of sklearn's [GLM](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.TweedieRegressor.html#sklearn.linear_model.TweedieRegressor)

### CODE REVIEW

File: linfa-linear/src/glm/distribution.rs [hunk 1]
```diff
@@ -4,6 +4,8 @@ use ndarray::Zip;
 
 use crate::error::{LinearError, Result};
 
+const NEG_INFINITY: f64 = -1.0_f64 / 0.0_f64;
```
Comment (paulkoerbitz) [inline linfa-linear/src/glm/distribution.rs:7]: I believe we should be using `std::ft64::NEG_INFINITY` instead (for compatibility with Rust 1.38). Better than defining our own constant.

Comment (paulkoerbitz) [reply linfa-linear/src/glm/distribution.rs:7]: (it's deprecated in favor of f64::NEG_INFINITY in new rust but not available in 1.38, so since we want to be backwards compatible we need to use that instead).

Comment (VasanthakumarV) [reply linfa-linear/src/glm/distribution.rs]: Thanks, I have made the change.

File: linfa-linear/src/glm.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+        let (ypred, _, offset) = self.ypred(&p);
+
+        let dev = self.dist.deviance(self.y, &ypred)?;
+
+        let pscaled = p
+            .slice(s![offset..])
+            .mapv(|x| x * A::from(self.alpha).unwrap());
+
+        let obj = A::from(0.5).unwrap() * dev
```
Comment (bytesnake) [inline linfa-linear/src/glm.rs:226]: can you add a comment here, where this comes from? You can leave out one of the `0.5` here :)

**Review by bytesnake — APPROVED** [review]
looks good to me, like the large number of tests :+1:

File: linfa-linear/src/glm.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+        let problem = TweedieProblem {
+            x,
+            y,
+            fit_intercept: self.fit_intercept,
+            link: &link,
+            dist,
+            alpha: self.alpha,
+        };
+        let linesearch = MoreThuenteLineSearch::new();
+        let solver = LBFGS::new(linesearch, 7).with_tol_grad(A::from(self.tol).unwrap());
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs:158]: excerpt from wiki on [LBFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS),
> L-BFGS maintains a history of the past m updates of the position x and gradient ∇f(x), where generally the history size m can be small (often m < 10)

`argmin` package doesn't explain the `m` variable, I took the `m` value used in their examples as a default.

**Comment by VasanthakumarV** [comment]
@bytesnake Thank you for the review. I will be pushing the changes at the earliest.

File: linfa-linear/src/glm.rs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+        if self.alpha < 0. {
+            return Err(LinearError::InvalidValue(format!(
+                "Penalty term must be a non-negative number, got: {}",
+                self.alpha
+            )));
+        }
+
+        let link: Link;
+        if let Some(value) = self.link {
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs:114]: `or_else` seems to return an `Option`, so returning a `Result` from the closure body is not possible. I have been trying out different functional style functions available in the [option doc](https://doc.rust-lang.org/std/option/enum.Option.html) but couldn't find the right combination.

File: linfa-linear/src/glm.rs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+/// | ------ | ---------------------- |
+/// | 0      | Normal                 |
+/// | 1      | Poisson                |
+/// | (1, 2) | Compound Poisson Gamma |
+/// | 2      | Gamma                  |
+/// | 3      | Inverse Gaussian       |
+///
+/// NOTE: No distribution exists between 0 and 1
+///
+/// Learn more from sklearn's excellant [User Guide](https://scikit-learn.org/stable/modules/linear_model.html#generalized-linear-regression)
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs]: Spelling corrected.

File: linfa-linear/src/glm.rs [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+            // for the distribution
+            return Err(LinearError::InvalidValue(format!(
+                "Some value(s) of y are out of the valid range for power value {}",
+                self.power
+            )));
+        }
+
+        // We initialize the coefficients and intercept
+        let mut coef: Array1<A>;
+        if self.fit_intercept {
```
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs:139]: This is definitely better, made the change.

File: linfa-linear/src/glm.rs (hunk 2 above)
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs]: I made the change with the multiplier.
I have added a comment on the `apply` function, which encapsulates the objective function we are trying to minimize using LBFGS.

File: linfa-linear/src/glm.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bytesnake) [reply linfa-linear/src/glm.rs:114]: ah damn :( you could combine match and guards then, but it's a bit harder to read:
```rust
let link = match self.link {
    Some(x) => Ok(x),
    None if self.power <= 0. => Ok(Link::Identity),
    None if self.power >= 1. => Ok(Link::Log),
    None => Err(LinearError::InvalidValue(format!("{}", self.power)))
}?;
```

File: linfa-linear/src/glm.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bytesnake) [reply linfa-linear/src/glm.rs:154]: okay that are the number of gradient samples from which the hessian is approximated, perhaps just add the wiki as a comment with m=7

[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs:158]: I have added the comment.

File: linfa-linear/src/glm.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (VasanthakumarV) [reply linfa-linear/src/glm.rs]: The alternative with the guards avoid one level of nesting, so I have made the change.

**Comment by bytesnake** [comment]
I'm ready to merge, but you still have to add GLM to the overview

**Comment by VasanthakumarV** [comment]
> I'm ready to merge, but you still have to add GLM to the overview

@bytesnake Can you point me to the overview section? Is this the [current state](https://github.com/rust-ml/linfa#current-state) section?

**Comment by bytesnake** [comment]
@VasanthakumarV yes the notes under notes add your algorithm

**Comment by VasanthakumarV** [comment]
@bytesnake I have added it, 
One clarification partial_fit means incremental learning right? 
GLM doesn't support partial_fit. So should I change something?

**Comment by bytesnake** [comment]
I think this refers to the algorithm itself, so probably just leave it there ..
